chore: update dependencies to resolve security vulnerabilities

ti-jmcevoy opened the pull request:
BREAKING CHANGE: Multiple major version updates that may affect consumers:
- Migrated from vite-plugin-ssr to vike (v0.4.195)
- Updated Storybook from v7 to v8.6.14
- Updated i18next from v21 to v23.16.8
- Updated react-i18next from v11 to v13.5.0

Security fixes (8 vulnerabilities resolved - 3 low, 5 moderate):
- graphql: 16.6.0 → 16.11.0 (uncontrolled resource consumption)
- universal-cookie: ^4.0.4 → ^7.2.2 (cookie vulnerability)
- remark-mdx-frontmatter: ^2.0.3 → ^5.2.0 (prototype pollution)
- tsup: various → ^8.5.0 (esbuild vulnerability)
- Removed unused patch-package (tmp vulnerability)

Package updates across all workspaces:
- vite: ^3.2.8 → ^5.4.11
- @vitejs/plugin-react: ^2.1.0 → ^4.3.4
- lerna: ^4.0.0 → ^8.1.9
- All Storybook packages updated to v8.6.14
- Consistent i18next versions across monorepo

Implementation changes:
- Fixed CJS/ESM compatibility with dynamic Vike imports
- Added deployment manifest generation scripts
- Updated Storybook Apollo addon for v8 compatibility
- Fixed React Hooks violations in i18next decorator
- Added Vite alias for NavBar import resolution
- Updated .gitignore with build artifacts and temp files

Affected packages:
- @thoughtindustries/helium-server
- @thoughtindustries/helium-template
- @thoughtindustries/helium-template-essentials
- @thoughtindustries/storybook-addon-apollo-client
- All @thoughtindustries component packages

ti-jmcevoy committed:
feat!: Major package updates, Vike migration, and routing fixes

BREAKING CHANGES:

1. Vite-plugin-ssr → Vike Migration
   - All imports changed from 'vite-plugin-ssr/*' to 'vike/*'
   - Package renamed: vite-plugin-ssr is now vike
   - Documentation URLs changed from vite-plugin-ssr.com to vike.dev

2. Storybook v7 → v8 Upgrade
   - @storybook/addons → @storybook/manager-api
   - @storybook/addon-docs now required explicitly
   - @storybook/addon-postcss removed (incompatible with v8)
   - storybook-addon-cookie v4 auto-registers (remove manual imports)
   - Custom addons must update to functional decorator API

3. Major Dependency Version Bumps
   - Vite: ^3.2.8 → ^5.4.11
   - lerna: ^4.0.0 → ^8.1.9
   - universal-cookie: ^4.0.4 → ^7.2.2 (API changes)
   - remark-mdx-frontmatter: ^2.0.3 → ^5.2.0 (API changes)
   - i18next: ^21.10.0 → ^23.16.8 (config changes)
   - react-i18next: ^11.18.6 → ^13.5.0
   - tsup: ^5.x/^6.x → ^8.5.0 (config changes)
   - @types/node: ^17.0.23 → ^18.19.68

4. Build Process Changes
   - New script: generate-manifest.js required after build
   - Build command: "vite build && node scripts/generate-manifest.js"
   - Manifest generation now automatic from Vike's assets.json
   - SSR bundling: Must add ESM packages to ssr.noExternal config

5. React Pattern Enforcement
   - Apollo onCompleted callbacks no longer used for state updates
   - All state updates moved to useEffect hooks
   - Components using old patterns will cause React errors #418/#423

FEATURES:

1. Improved Client Routing Support
   - Vike always generates both entry-server-routing.js and entry-client-routing.js
   - Automatic routing mode detection from server bundle
   - Fixed asset generation to include correct entry file
   - Fixed Content-Type headers for .pageContext.json requests
   - Dev server now properly handles Client Routing navigation

2. Hybrid SSR Implementation
   - Server Routing: Always SSR (traditional behavior)
   - Client Routing: SSR on initial load, client-side on navigation
   - Automatic SSR mode detection in catalog components
   - Best of both worlds: SEO + smooth navigation

3. React Hydration Improvements
   - Added isHydrated flag to prevent multiple hydration calls
   - Proper root management for Client Routing
   - Fixed hydration errors in production

4. Client-Side Data Fetching
   - Added useQuery hooks for Client Routing navigation
   - Course detail page fetches data on navigation
   - Component remounting with key props for data refresh

FIXES:

1. Critical React Errors Fixed
   - Fixed LoadBookmarks setState in render body
   - Fixed LearnerAccess onCompleted state updates
   - Moved all state updates to useEffect hooks
   - Fixes React errors #418 (update during render) and #423 (update during transition)

2. Content-Type Headers Fixed
   - Production worker now returns application/json for .pageContext.json
   - Dev server returns JSON for Client Routing requests
   - Fixed "Wrong Content-Type" errors during navigation

3. SSR Build Fixed
   - Cannot import client-only files (_default.page.client.tsx) on server
   - Use pageContext.isHydration instead for render detection
   - Fixed Vike file-env restriction violations

4. Asset Generation Fixed
   - Automatically detect routing mode from dist/server/entry.mjs
   - Include correct entry file based on detected mode
   - Both routing modes now work correctly

5. TypeScript Fixes
   - Reordered package.json exports (types first)
   - Fixed "types condition should be first" warnings

SECURITY:

- Reduced vulnerabilities from 11 (3 low, 8 moderate) to 3 (0 low, 3 moderate)
- Fixed security issues in:
  - lerna (outdated version)
  - universal-cookie <0.7.0
  - graphql <16.11.0 (uncontrolled resource consumption)
  - remark-mdx-frontmatter (prototype pollution)
  - Storybook packages (multiple moderate issues)
  - esbuild via tsup upgrade

Remaining 3 moderate vulnerabilities are esbuild <=0.24.2 but only affect Vite dev server

PACKAGE UPDATES:

Root Package (helium/package.json):
- lerna: ^4.0.0 → ^8.1.9
- autoprefixer: ^10.3.6 → ^10.4.20
- postcss: ^8.3.8 → ^8.4.49
- All @storybook/*: ^7.0.21 → ^8.6.14
- storybook-addon-cookie: ^3.0.0 → ^4.0.1
- storybook-react-i18next: ^1.1.2 → ^3.3.1
- i18next: ^21.10.0 → ^23.16.8
- react-i18next: ^11.18.6 → ^13.5.0
- vite: NEW ^5.4.11 (required by Storybook v8)
- patch-package: REMOVED (unused, 2 vulnerabilities)

Helium Server (tooling/helium-server/package.json):
- vite: ^3.2.8 → ^5.4.11
- vite-plugin-ssr: ^0.4.43 → REMOVED
- vike: NEW ^0.4.195
- graphql: ^16.6.0 → ^16.11.0
- @vitejs/plugin-react: ^2.1.0 → ^4.3.4
- tsup: ^5.11.13 → ^8.5.0
- express: ^4.18.2 → ^4.21.2

Template Packages (tooling/template-base*/package.json):
- vite: ^3.2.8 → ^5.4.11
- vite-plugin-ssr: 0.4.141 → REMOVED
- vike: NEW 0.4.195
- graphql: 16.6.0 → 16.11.0
- remark-mdx-frontmatter: ^2.0.3 → ^5.2.0
- tsup: ^6.2.1 → ^8.5.0
- @types/node: ^17.0.23 → ^18.19.68

Cart Package (packages/cart/package.json):
- universal-cookie: ^4.0.4 → ^7.2.2

Total Changes:
- 35 packages upgraded
- 8 new packages added
- 5 packages removed
- 8 critical bug fixes
- 16 files modified with routing fixes
- ~500 insertions, ~200 deletions

rjschill87 commented on tooling/template-base/vite.config.ts:
```diff
@@ -1,7 +1,20 @@
+
+  try {
+    // Try to load vike dynamically
+    const vikeModule = await import('vike/plugin');
```
Is there a reason to load vike dynamically here vs importing it from the start? Seems like it could fail silently and SSR would break -- so it'd look like it was working, when it's really limping since it'd be hard to see with it just warning. Feels like something we should surface?

ti-jmcevoy replied:
New Implementation:

```node
// In helium-server/src/vite-config/vite.config.ts
export async function createVikeConfig() {
  try {
    const vikeModule = await import('vike/plugin');
    const vike = vikeModule.default || vikeModule;
    
    return mergeConfig(baseConfig, {
      plugins: [vike()]
    });
  } catch (error) {
    console.error('Failed to load vike plugin:', error);
    throw new Error('Vike plugin is required but could not be loaded. Please ensure vike is installed.');
  }
}
```

**Why dynamic import:** We encountered ESM/CJS compatibility issues where Node.js was trying to `require()` an ESM-only module (vike), causing "ERR_REQUIRE_ESM" errors

**Addresses silent failure concern:** We throw an explicit error that stops the build/dev server if vike fails to load

**Clear error messaging:** The error message clearly states what's wrong and what's needed

**No Partial Functionality:** The application won't start without vike, ensuring no partial functionality

These ESM/CJS compatibility issues are really tricky and finicky especially with Vike so this was the best solution I could come up with that didn't break everything.

rjschill87 commented on tooling/template-base/vite.config.ts:
```diff
@@ -10,7 +23,29 @@ export default defineConfig(async () => {
-  return tiConfig;
+  return {
+    plugins,
+    build: {
```
Reason to add these `build` and other props here? Seems like the thing we'd cover in the `vote.config.ts` that ships with `helium-server`? Seems like we could end with issues where someone turns off `manifest`, or changes the `envPrefix` (which I'm pretty sure we wouldn't expect right now) and cause issues?

ti-jmcevoy replied:
New Implementation:

```node
// In helium-server/src/vite-config/vite.config.ts
export const baseConfig = {
  build: {
    manifest: true,  // Required for production asset mapping
  },
  ssr: {
    noExternal: ['@apollo/client', 'graphql', '@thoughtindustries/helium-*']
  },
  envPrefix: 'VITE_',
  resolve: {
    dedupe: ['graphql', '@apollo/client']
  }
};

// Templates now just extend this:
// In template-base/vite.config.ts
const config = await createVikeConfig();
config.plugins.push(mdx());
```

**Centralized control:** All critical settings are now in `helium-server`, templates can't accidentally break them

**Prevents misconfiguration:** Templates can't turn off `manifest: true` which would break production deployments

**Consistent envPrefix:** Ensures environment variables work consistently across all Helium apps

**Extensible:** Templates can still add their own plugins and settings without overriding critical ones

rjschill87 commented on tooling/helium-server/src/server/server.ts:
```diff
@@ -1,12 +1,31 @@
 import path from 'path';
 
+// Lazy load vike/server to handle CJS/ESM compatibility issues
+let vikeModule: any = null;
```
Think we may want to investigate a different approach here, or at least do some load testing; I'm not sure how Workers handles concurrent requests, but there's potential for a race condition here if it's allowed. Multiple concurrent requests mean multiple imports of `vike/server`.

Is there an issue w/ importing it from the start regardless of env, and instead only overwrite it in dev env?

ti-jmcevoy replied:
New Implementation:

```node
// In helium-server/src/server/server.ts
let renderPage: any = null;  // Module-level cache

export async function setupHeliumServer(...) {
  // Load renderPage dynamically to avoid CJS/ESM issues
  if (!isProduction && viteDevServer) {
    // In development, use dev server's module loader for HMR support
    const vikeDevModule = await viteDevServer.ssrLoadModule('vike/server');
    renderPage = vikeDevModule.renderPage;
  } else if (!renderPage) {  // Only import once in production
    // In production or first time, dynamically import vike
    const vikeModule = await import('vike/server');
    renderPage = vikeModule.renderPage;
  }
}
```

**Race condition prevented:** The `!renderPage` check ensures import happens only once

**Module-level caching:** Once loaded, `renderPage` is cached for all subsequent requests

**Different approach for dev:** Uses Vite's SSR module loader in development for HMR support

**Production optimization:** In production, the import happens once during server startup, not per request

**Worker consideration:** Cloudflare Workers are single-threaded within an isolate, but we still cache to avoid redundant work

rjschill87 commented on tooling/template-base/renderer/_default.page.client.tsx:
```diff
@@ -66,4 +81,18 @@ async function render(pageContext: PageContext) {
+    isHydrated = true;
+  } else {
+    // Client-side navigation - reuse the root for all subsequent renders
+    if (!root) {
```
It predates your changes, but think it might be worth implementing the same pattern you've done for `root` for the `apolloClient` -- e.g., creating a `currentApolloClient` so that we can stop/clear the previous Apollo client so we don't get any memory leaks. Something like:

```javascript
if (currentApolloClient) {
  currentApolloClient.stop();
  currentApolloClient.clearStore();
}

currentApolloClient = apolloClient

```

ti-jmcevoy replied:
New Implementation:

```node
// Track current Apollo Client for proper cleanup
let currentApolloClient: ReturnType<typeof makeApolloClient> | null = null;

async function render(pageContext: PageContext) {
  // ... extract pageContext properties

  // Clean up previous Apollo Client to prevent memory leaks
  if (currentApolloClient) {
    // Stop all active queries and subscriptions
    currentApolloClient.stop();
    // Clear the store to free up memory
    await currentApolloClient.clearStore();
    currentApolloClient = null;
  }

  // Create Apollo Client - always create fresh on client side
  const apolloClient = await makeApolloClient(
    heliumEndpoint,
    apolloInitialState,
    isProduction,
    authToken
  );
  
  // Track the current client
  currentApolloClient = apolloClient;
  
  // ... rest of render logic
}
```

**Addresses the Core Concern:** Prevents memory leaks by explicitly cleaning up old Apollo Client instances

**Proper Cleanup Sequence:**
- `stop()` > Terminates all active queries and subscriptions
- `clearStore()` >  Frees cache memory
- `null assignment` > Allows garbage collection

**Applied Consistently:** Implemented in both template-base and template-base-essentials

**Avoids Vike Errors:** Doesn't try to access `pageContext.apolloClient` (which isn't passed to client)

**Performance Impact:** Prevents memory growth in long-running sessions with many navigations

rjschill87 commented on tooling/helium-server/worker/static-assets.js:
```diff
@@ -15,8 +19,90 @@ export { handleStaticAssets };
+      // Try to match by removing the extension and looking for similar files
+      const pathWithoutExt = cleanPath.replace(/\.(js|css)$/, '');
+
+      for (const [originalPath, hashedPath] of Object.entries(manifest)) {
```
Did this come from CF or Vice? I'm wondering if there's another way we can implement this -- iterating through the entire manifest on every request seems... expensive. Would it work if we just build a map once when the worker starts? Might want to tighten up the hash matching also, it seems pretty fuzzy.

ti-jmcevoy replied:
New Implementation:

```node
// In helium-server/worker/ssr.js
// Parse asset manifest once at worker startup for better performance
// Cloudflare Workers are long-lived, so this cache persists across requests
let CACHED_ASSET_URLS = null;
if (typeof __STATIC_CONTENT_MANIFEST !== 'undefined') {
  try {
    const assetManifest = JSON.parse(__STATIC_CONTENT_MANIFEST);
    CACHED_ASSET_URLS = getAssetUrls(assetManifest);
    console.log('Asset manifest parsed and cached at worker startup');
  } catch (error) {
    console.error('Failed to parse asset manifest at startup:', error);
  }
}

async function handleSsr(url, authToken = null, userAndAppearanceToken = null) {
  // ... 
  const pageContext = await initPageContext(
    url,
    renderPage,
    currentUser,
    appearanceBlock,
    HELIUM_ENDPOINT,
    true,
    sha256,
    authToken,
    null,
    CACHED_ASSET_URLS // Use pre-parsed asset URLs from startup
  );
}
```

**One-time parsing:** Manifest is parsed once when the Worker starts, not on every request

**Long-lived Workers:** Cloudflare Workers persist between requests within an isolate, so the cache remains valid

**Performance improvement:** Eliminates O(n) iteration through manifest on each request

**Memory efficient:** Single shared cache across all requests in the Worker isolate

**Fail-safe:** If parsing fails at startup, it logs the error but doesn't crash the Worker

rjschill87 reviewed (commented):
Looks good so far! Still making my way through, but left a few comments on a couple things. Appreciate all the work here!

ti-jmcevoy committed:
fix: resolve Vike SSR issues and address code review feedback

This commit addresses and implements reviewer suggestions, and fixes
production deployment issues.

## Core Fixes

### 1. Vike Worker Initialization (Production Fix)
- Import dist/server/entry.mjs before vike/server in worker
- Resolves '@brillout/vite-plugin-server-entry' missing error
- Ensures Vike is properly initialized with built application code

### 2. ESM/CJS Compatibility Resolution
- Created createVikeConfig() factory function with dynamic imports
- Added --external vike to tsup build to prevent bundling
- Pinned vike to exact version 0.4.195 for consistency
- Dynamic imports in server.ts for dev/production environments

## Code Review Implementations

### 1. Centralized Vite Configuration
- Moved all critical build settings to baseConfig in helium-server
- Templates now extend config via createVikeConfig() factory
- Prevents accidental misconfiguration (manifest, envPrefix, etc.)

### 2. Race Condition Prevention
- Module-level caching of renderPage to prevent concurrent imports
- Import once, cache for all subsequent requests

### 3. Asset Manifest Performance
- Parse manifest once at worker startup
- Cache in CACHED_ASSET_URLS for all requests
- Eliminates O(n) iteration on each request

## Dependencies
- vike: Pinned to 0.4.195 (from ^0.4.195)

## Testing
- Verified Server Routing mode works correctly
- Verified Client Routing mode works correctly
- Tested production deployments on Cloudflare Workers

ti-jmcevoy committed:
fix: implement Apollo Client cleanup to prevent memory leaks

Addresses reviewer's concern about potential memory leaks in long-running
application sessions with multiple navigations.

Changes:
- Track currentApolloClient at module level
- Clean up previous client before creating new one:
  - Call client.stop() to stop all active queries/subscriptions
  - Call client.clearStore() to free cache memory
  - Set to null to allow garbage collection
- Always create fresh Apollo Client on client side (no reuse from SSR)
- Applied to both template-base and template-base-essentials

Fixed issue with pageContext.apolloClient access:
- Removed check against pageContext.apolloClient (not in passToClient)
- Client always creates its own Apollo Client using apolloInitialState

This prevents:
- Memory growth from accumulating Apollo Client instances
- Active subscriptions/queries from old clients continuing to run
- Cache data from previous navigations being retained unnecessarily
- Vike errors about accessing non-passed properties

In long-running sessions with many navigations, this ensures proper
cleanup and prevents performance degradation over time.

rjschill87 commented on tooling/helium-server/src/vite-config/vite.config.ts:
```diff
@@ -1,31 +1,51 @@
-    baseConfig.plugins.push(vike());
+
+    return {
+      ...baseConfig,
```
May want to double check me but think we'd want to use a deep clone here instead of a spread. Nested objects will still be a reference to the original, so if file A calls `createVikeConfig` and makes a change to `resolve`, then when file B calls `createVikeConfig` later it'll get the version w/ the altered `resolve`.

ti-jmcevoy replied:
Good catch, I've resolved this by using structuredClone() to deep clone all serializable config properties. Here is the new implementation:

```node
// Factory function to create config with vike plugin
// Uses dynamic import to avoid ESM/CJS compatibility issues
// Templates call this to get the complete config
export async function createVikeConfig() {
  try {
    const vikeModule = await import('vike/plugin');
    const vike = vikeModule.default || vikeModule;

    // Deep clone baseConfig to prevent shared references between consumers
    // Note: We can't clone plugins (they contain functions), so we handle them separately
    // This ensures that if one consumer modifies nested objects (e.g., resolve.alias),
    // it won't affect other consumers or the original baseConfig
    const { plugins, ...serializableConfig } = baseConfig;
    const clonedConfig = structuredClone(serializableConfig);

    return {
      ...clonedConfig,
      plugins: [...plugins, vike()] // Use original plugins array (safe - replaced not mutated)
    };
  } catch (error) {
```

rjschill87 commented on tooling/helium-server/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@thoughtindustries/helium-server",
-  "version": "3.0.0-beta.15",
+  "version": "3.0.0-beta.24",
```
More of an all-over thing (or mostly tooling, since components are mostly just react), but w/ all the major changes we might want to update the major version so any folks w/ their dependencies not properly pinned aren't affected.

ti-jmcevoy replied:
The current released non-beta version for helium-server is `2.7.0`, at the moment while testing I have been publishing beta packages to NPM so that they can be installed by the helium-deployer for staging instances. I haven't yet published a new production package, but once I do (once this review is complete) I will publish new production packages to NPM with a new major version `3.0.0` due to all of the breaking/large changes and this will be reflected in the package.json file.

rjschill87 replied:
@ti-jmcevoy Not quite following, can you clarify? The `helium-deployer` should still be able to install a major change beta fine (for staging at least) without worrying about production at all. Looks like it treats any `Vite` dependency `>=3` as a Worker V2, e.g., it goes off the `Vite` version, not our package version.

ti-jmcevoy replied:
Apologies, I probably wasn't super clear with the above.

Currently in `staging`, `tooling/helium-server/package.json` is pinned to the currently published version:
https://github.com/thoughtindustries/helium/blob/staging/tooling/helium-server/package.json#L3

At the moment this is version `2.7`

When publishing the beta versions to NPM lerna automatically adjusted this for me to the latest beta version that we were currently publishing, which makes sense as the published beta package to NPM should have it's own version properly defined here, I committed this like this because we haven't actually published the final version yet, my plan was to let lerna automatically change this to `3.0.0` once I publish the final package after this review has been approved and then pushing those changes to the branch, but in this case seeing as we know that the final version will be `3.0.0` I can just update this to the final version prior to publishing the package with lerna which should resolve this

ti-jmcevoy replied:
This is now updated with lerna to the correct versions and ready to publish to npm

rjschill87 replied:
@ti-jmcevoy -- Looks like doing the major vers change dropped the beta flag? We'll still need that since we'll need to publish packages to test in staging. Been a while but think you might be able to do it with something like 

```sh
lerna version major --conventional-prerelease --message "chore: release new versions" --include-merged-tags
```

But it's been a while so I'd def re-check the docs.

Last question too -- were there any change in the `env` var prefixes that vite/vike uses now? Only asking so they still get picked up by the deployer: 

https://github.com/thoughtindustries/iron-workers/blob/c4f81332c4c17e9a2c26b1e3b1d17425bb321a78/helium-deployer/parse-env-vars.js

ti-jmcevoy committed:
fix: use structuredClone for deep cloning Vite config

Addresses reviewer's concern about shared references in nested objects
when multiple consumers call createVikeConfig().

Changes:
- Separate plugins from serializable config (plugins contain functions)
- Use structuredClone() on serializable parts (build, optimizeDeps, ssr, resolve)
- Reconstruct config with cloned serializable parts + fresh plugins array

This prevents mutations to nested objects (e.g., resolve.alias) in one
consumer from affecting other consumers or the original baseConfig.

Note: Plugins cannot be cloned (contain functions), but they are already
handled safely since consumers replace the plugins array rather than
mutating individual plugins.

Reviewer: @rjschill87

ti-jmcevoy committed:
fix: update Node.js requirement to 18.0.0+ for Helium 3.0

Vike 0.4.195 and Vite 5 both require Node 18+, so the engines field
must reflect this requirement.

Updated engines.node in:
- @thoughtindustries/helium-server: >=16.18.1 → >=18.0.0
- @thoughtindustries/helium-template: >=16.18.1 → >=18.0.0
- @thoughtindustries/helium-template-essentials: >=16.18.1 → >=18.0.0

ti-jmcevoy committed:
chore: release 3.0.0 beta versions

 - @thoughtindustries/cart@2.0.0-beta.2
 - @thoughtindustries/catalog@2.0.0-beta.2
 - @thoughtindustries/content@2.0.0-beta.2
 - @thoughtindustries/dashboard-stats@2.0.0-beta.2
 - @thoughtindustries/featured-content@2.0.0-beta.2
 - @thoughtindustries/link-lists@2.0.0-beta.2
 - @thoughtindustries/navigation-bar@2.0.0-beta.2
 - @thoughtindustries/user@2.0.0-beta.2
 - @thoughtindustries/video-player@2.0.0-beta.2
 - create-helium-app@2.0.0-beta.30
 - @thoughtindustries/graphql-codegen-plugin@2.0.0-beta.2
 - @thoughtindustries/helium-server@3.0.0-beta.25
 - @thoughtindustries/storybook-addon-apollo-client@2.0.0-beta.2
 - @thoughtindustries/helium-tailwind-preset@2.0.0-beta.2
 - @thoughtindustries/helium-template@2.0.0-beta.35
 - @thoughtindustries/helium-template-essentials@2.0.0-beta.28

ti-jmcevoy committed:
fix: merge staging, resolve conflicts keeping Apollo Client approach
